src/transaction_processor.py

```diff
@@ -0,0 +1,182 @@
+    ):
+        """Function processes a single tx to find imbalances, prices."""
+        token_imbalances = self.imbalances.compute_imbalances(tx_hash)
+        if token_imbalances is not None:
+            self.log_message.clear()  # reset log message for each transaction
+            self.log_message.append(
+                f"Token Imbalances on {self.chain_name} for tx {tx_hash}:"
+            )
+            for token_address, imbalance in token_imbalances.items():
+                if imbalance != 0:
+                    self.write_token_imbalances(
+                        tx_hash, auction_id, block_number, token_address, imbalance
+                    )
+                    self.write_prices_if_available(tx_hash, block_number, token_address)
+
+                    self.log_message.append(
+                        f"Token: {token_address}, Imbalance: {imbalance}"
+                    )
+
+            logger.info("\n".join(self.log_message))
+        else:
+            raise ValueError("Imbalances computation returned None.")
+
+    def record_exists(self, tx_hash: bytes, token_address: bytes) -> bool:
+        """
+        Function checks if an entry of (tx_hash, token) already exists in the token imbalances table.
+        """
+        query = read_sql_file("src/sql/select_record_exists.sql")
+        result = self.db.execute_query(
+            query, {"tx_hash": tx_hash, "token_address": token_address}
+        )
+        return result.fetchone() is not None
+
+    def write_token_imbalances(
+        self,
+        tx_hash: str,
+        auction_id: int,
+        block_number: int,
+        token_address: str,
+        imbalance: float,
+    ):
+        """Function attempts to write token imbalances to the table."""
+        tx_hash_bytes = bytes.fromhex(tx_hash[2:])
+        token_address_bytes = bytes.fromhex(token_address[2:])
+
+        if not self.record_exists(tx_hash_bytes, token_address_bytes):
+            query = read_sql_file("src/sql/insert_raw_token_imbalances.sql")
+            self.db.execute_and_commit(
+                query,
+                {
+                    "auction_id": auction_id,
+                    "chain_name": self.chain_name,
+                    "block_number": block_number,
+                    "tx_hash": tx_hash_bytes,
+                    "token_address": token_address_bytes,
+                    "imbalance": imbalance,
+                },
+            )
+
+    def write_prices_if_available(
+        self, tx_hash: str, block_number: int, token_address: str
+    ):
+        """Function checks if price data can be fetched for writing to table."""
+        price_data = self.price_providers.get_price(block_number, token_address)
+        if price_data:
+            price, source = price_data
+            self.write_prices(source, block_number, tx_hash, token_address, price)
+            self.log_message.append(f"Token: {token_address}, Price: {price} ETH")
```

Comment:
This is where the attribute `self.log_message` is used, I suppose. I would probably restructure the main loop to make this necessary.